File: rag/advanced_rag/harness/memory.py
"""Retrieval memory — a central store of every raw chunk any claim has retrieved.

Purpose
-------
Throughout a multi-hop answer the system retrieves many chunks (search is cheap)
but hands the LLM only a small narrowed slice (LLM calls are slow & token-heavy).
The raw chunks must not be thrown away: the LLM's per-claim report/grounded
extraction can compress away a fact the answer needs (e.g. a candidate set
"R, RMP, RA, RF, RIF" collapsed to "RIF"). ``RetrievalMemory`` keeps ALL raw
chunks that any search returned, so the finalizer can do a cheap, deterministic,
no-LLM, no-knowledge-base relevance search over the already-retrieved corpus to
recover the missing fact instead of re-querying.

Design (in-memory, language-agnostic)
-------------------------------------
- ``add`` stores every raw chunk exactly as returned by the search backend,
  de-duplicated by chunk identity, BEFORE any narrowing — it is the lossless
  store backing the (lossy) ``kbinfos["chunks"]`` that feeds the LLM.
- ``search`` is the primary consumer-facing primitive: a relevance-ranked lookup
  over memory for a query string, language-agnostic (Latin words + numbers + CJK
  character 3-grams), requiring a normalized overlap bar so Chinese and English
  queries behave alike. It never invokes the LLM and never triggers a knowledge-
  base search.
- ``grep`` is a lower-level loose-keyword primitive (word-boundary + prefix
  tolerance) used as a building block / for tests; production finalize uses the
  stricter ``search``.

The store lives on ``tools.kbinfos["memory"]`` so it is serialized / carried with
the rest of the research state and is visible to any tool that has ``tools``.
"""

import logging
import re

_LOG = logging.getLogger(__name__)

# How many memory chunks a single grep query may return at most (keeps the
# injected evidence bounded).
_GREP_MAX_CHUNKS = 6
# Max sentences kept per chunk (hit + context) so a long chunk does not blow up
# the injection.
_GREP_MAX_SENTENCES = 4
# Absolute char budget per side of a hit when expanding context.
_GREP_CONTEXT_CHARS = 400
# Short chunks are kept whole (answers often live in short chunks).
_SHORT_CHUNK_CHARS = 200


def _chunk_key(ck: dict) -> str:
    return str(ck.get("chunk_id") or ck.get("id") or id(ck))


def _chunk_text(ck: dict) -> str:
    """The searchable text of a chunk, preferring the raw original."""
    for k in ("content", "content_with_weight"):
        v = ck.get(k)
        if v:
            return str(v)
    return ""


def _escape_term(term: str) -> str:
    t = str(term).strip()
    if not t:
        return ""
    t = re.sub(r"^[\s.,:;!?'\"()\[\]{}]+|[\s.,:;!?'\"()\[\]{}]+$", "", t)
    if not t:
        return ""
    escaped = re.escape(t)
    # CJK: no \b (Python re \b is ASCII-only and would never match).
    if re.search(r"[\u4e00-\u9fff\u3040-\u30ff\uac00-\ud7af]", t):
        return escaped
    if len(t) >= 3 and t[0].isalnum() and t[-1].isalnum():
        return rf"\b{escaped}\b"
    return escaped


def _split_sentences(text: str) -> list[str]:
    """Split into sentences, treating block HTML/markdown tables as atomic."""
    # Reuse the shared sentence splitter from tools.search when available, else a
    # minimal one. Import lazily to avoid a heavy module graph at load time.
    try:
        from rag.advanced_rag.harness.tools.search import _split_sentences as _ss

        return _ss(text)
    except Exception:
        pass
    return [s for s in re.split(r"(?<=[.!?。！？])\s+", (text or "").strip()) if s]


def _sentence_span_window(sents: list[str], idx: int) -> list[str]:
    """The hit sentence plus up to one neighbouring sentence on each side."""
    lo = max(0, idx - 1)
    hi = min(len(sents), idx + 2)
    window = sents[lo:hi]
    # Clamp total length.
    total = 0
    kept = []
    for s in window:
        total += len(s)
        if total > _GREP_CONTEXT_CHARS * 2:
            break
        kept.append(s)
    return kept or [sents[idx]]


def add(tools, chunks) -> None:
    """Merge raw retrieved chunks into the central memory store (lossless)."""
    if not chunks:
        return
    mem = tools.kbinfos.setdefault("memory", [])
    seen = {_chunk_key(c) for c in mem}
    added = 0
    for c in chunks:
        if not isinstance(c, dict) or not _chunk_text(c):
            continue
        k = _chunk_key(c)
        if k in seen:
            continue
        seen.add(k)
        mem.append(c)
        added += 1
    if added:
        _LOG.info("[Memory] stored %d new raw chunk(s); memory now has %d.", added, len(mem))


def grep(tools, terms, limit: int = _GREP_MAX_CHUNKS) -> list[dict]:
    """Return memory chunks that contain any of ``terms`` (word-boundary grep),
    narrowed to the matching sentence + small context.

    ``terms`` is a list of plain strings (entities / numbers / key phrases) as
    emitted by the analysis LLM. Returns a list of chunk dicts each carrying a
    narrowed ``content`` (a plain string) so the caller can splice them straight
    into an evidence list. Empty on no-hit / no-memory.
    """
    mem = tools.kbinfos.get("memory", []) or []
    if not mem or not terms:
        return []
    patterns = []
    # Prefix fallback: morphological tolerance. A gap term often differs from the
    # chunk's word by a suffix/prefix (gap "abbreviation" vs chunk "abbreviated";
    # gap "rifampin" vs chunk "Rifampicin"). We match the term's leading stem
    # (first 5 chars for terms >= 6) at the START of a chunk word, so a shared
    # root still hits without a short token over-matching. CJK terms are matched
    # verbatim (no stemming semantics).
    prefix_patterns = []
    for t in terms:
        frag = _escape_term(t)
        if frag:
            try:
                patterns.append(re.compile(frag, re.IGNORECASE))
            except re.error:
                continue
        _stripped = str(t).strip()
        _prefix = _stripped[:5] if len(_stripped) >= 6 else ""
        if _prefix and not re.search(r"[\u4e00-\u9fff\u3040-\u30ff\uac00-\ud7af]", _prefix):
            try:
                prefix_patterns.append(re.compile(rf"\b{re.escape(_prefix)}", re.IGNORECASE))
            except re.error:
                continue
    if not patterns and not prefix_patterns:
        return []

    def _match(text: str) -> bool:
        if any(p.search(text) for p in patterns):
            return True
        # Prefix fallback: the term's leading stem occurs at the START of a
        # chunk word (word boundary), tolerating inflectional suffixes.
        if prefix_patterns:
            for pp in prefix_patterns:
                if pp.search(text):
                    return True
        return False

    hits = []
    for c in mem:
        text = _chunk_text(c)
        if len(text) <= _SHORT_CHUNK_CHARS:
            # Short chunk: keep whole (its answer may live anywhere in it).
            if _match(text):
                hits.append({"content": text, "doc_id": c.get("doc_id"), "chunk_id": c.get("chunk_id")})
            continue
        sents = _split_sentences(text)
        kept = []
        for i, s in enumerate(sents):
            if _match(s):
                for w in _sentence_span_window(sents, i):
                    if w not in kept:
                        kept.append(w)
            if len(kept) >= _GREP_MAX_SENTENCES:
                break
        if kept:
            hits.append({"content": "\n".join(kept), "doc_id": c.get("doc_id"), "chunk_id": c.get("chunk_id")})
        if len(hits) >= limit:
            break
    return hits


def size(tools) -> int:
    return len(tools.kbinfos.get("memory", []) or [])


def clear(tools) -> None:
    tools.kbinfos["memory"] = []


# ─────────────────────────────────────────────────────────────────────────────
# Relevance-ranked retrieval over memory (used as a retrieval-reuse cache, NOT
# as a noise-injection source). Unlike ``grep`` (loose keyword hit), ``search``
# scores each memory chunk by how many SIGNIFICANT query terms it actually
# contains and returns only the chunks that clear a multi-term overlap bar —
# quality comparable to a knowledge-base retrieval, but served from memory so a
# fact retrieved earlier can be reused instead of re-querying the index.
# ─────────────────────────────────────────────────────────────────────────────

_STOPWORDS = {
    "what",
    "which",
    "how",
    "many",
    "much",
    "does",
    "did",
    "do",
    "the",
    "a",
    "an",
    "is",
    "are",
    "was",
    "were",
    "be",
    "been",
    "being",
    "of",
    "for",
    "to",
    "in",
    "on",
    "with",
    "and",
    "or",
    "by",
    "from",
    "at",
    "it",
    "its",
    "this",
    "that",
    "these",
    "those",
    "who",
    "when",
    "where",
    "why",
    "than",
    "then",
    "there",
    "their",
    "they",
    "them",
    "his",
    "her",
    "him",
    "she",
    "he",
    "we",
    "you",
    "your",
}

# CJK / other scripts have no word boundaries and no shared stopword list; match
# them as literal substrings. Script ranges: CJK unified ideographs, Hiragana,
# Katakana, Hangul, CJK punctuation is excluded.
_CJK_RE = re.compile(r"[\u4e00-\u9fff\u3040-\u30ff\uac00-\ud7af]+")
_LATIN_NUM_RE = re.compile(r"[A-Za-z0-9]+")


def _is_cjk(s: str) -> bool:
    return bool(_CJK_RE.fullmatch(s))


def _significant_terms(text: str, max_terms: int = 18) -> list[str]:
    """Language-agnostic significant-term extraction.

    - Latin alphanumeric runs → lowercased words (stopword-filtered, len >= 3).
    - CJK runs (Chinese/Japanese/Korean) → split into character 3-grams (no word
      boundaries exist, so n-grams are the language-agnostic way to match a
      substring like 利福平 inside a chunk); stopwords are NOT applied.
    - standalone numbers kept verbatim.
    De-duplicated, capped. Returns [] when nothing searchable.
    """
    out: list[str] = []
    seen: set[str] = set()

    def _push(tok: str) -> None:
        if tok and tok not in seen:
            seen.add(tok)
            out.append(tok)

    # Numbers anywhere.
    for m in re.finditer(r"\d+", text or ""):
        _push(m.group(0))
        if len(out) >= max_terms:
            return out
    # CJK runs → 3-grams (and the whole run if shorter than 3).
    for m in _CJK_RE.finditer(text or ""):
        run = m.group(0)
        if len(run) < 3:
            _push(run)
        else:
            for i in range(len(run) - 2):
                _push(run[i : i + 3])
        if len(out) >= max_terms:
            return out
    # Latin words (stopword-filtered).
    for m in _LATIN_NUM_RE.finditer(text or ""):
        raw = m.group(0)
        if raw.isdigit():
            continue  # numbers already handled
        low = raw.lower()
        if len(low) >= 3 and low not in _STOPWORDS:
            _push(low)
        if len(out) >= max_terms:
            return out
    return out


def _term_hits(text: str, terms: list[str]) -> int:
    """How many of ``terms`` occur in ``text``, language-agnostically.

    - CJK terms → literal substring (no word boundary).
    - Latin terms → word-boundary match with a short prefix fallback for
      inflectional variants ("abbreviation" also matches "abbreviated").
    - numbers → literal presence.
    """
    if not terms:
        return 0
    hits = 0
    for t in terms:
        if _is_cjk(t):
            if t in text:
                hits += 1
        elif t.isdigit():
            if t in text:
                hits += 1
        else:
            if re.search(rf"\b{re.escape(t)}\b", text, re.IGNORECASE):
                hits += 1
            elif len(t) >= 6:
                _prefix = re.escape(t[:5])
                if re.search(rf"\b{_prefix}", text, re.IGNORECASE):
                    hits += 1
    return hits


def search(tools, query: str, top_n: int = 6, min_overlap: int = 2, min_ratio: float = 0.12) -> list[dict]:
    """Relevance-ranked retrieval over memory for a query string (language-agnostic).

    Scores each memory chunk by how many of the query's significant terms
    (Latin words / numbers / CJK 3-grams) it matches, and keeps chunks that clear
    a *normalized* overlap bar so Chinese and English queries behave alike: a
    chunk is relevant when it shares ``>= 1`` term AND ``>= min_ratio`` of the
    query's terms (a long CJK query matches a short "利福平" via shared 3-grams;
    an English query matches "abbreviation → abbreviated" via word-boundary +
    prefix). Ranked by hit count. Cheap, deterministic, no-LLM — used to REUSE
    evidence already retrieved, never to inject loose noise. Returns [] when
    nothing clears the bar (caller falls back to the knowledge-base search).
    """
    mem = tools.kbinfos.get("memory", []) or []
    terms = _significant_terms(query)
    if not mem or not terms:
        return []
    _n = len(terms)
    scored = []
    for c in mem:
        text = _chunk_text(c)
        if not text:
            continue
        hits = _term_hits(text, terms)
        if hits >= 1 and (hits / _n) >= min_ratio:
            scored.append((hits, text, c))
    if not scored:
        return []
    scored.sort(key=lambda x: (-x[0], -len(x[1])))
    out = []
    for hits, text, c in scored[:top_n]:
        out.append({"content": text, "doc_id": c.get("doc_id"), "chunk_id": c.get("chunk_id"), "similarity": float(hits)})
    _LOG.info("[Memory.search] query=%r -> %d relevant chunk(s) (ratio>=%.2f, %d terms)", (query or "")[:60], len(out), min_ratio, _n)
    return out

File: internal/rag/advanced_rag/harness/memory.go
//
//  Copyright 2026 The InfiniFlow Authors. All Rights Reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

package harness

import (
	"regexp"
	"sort"
	"strings"
	"unicode"
)

// Raw-chunk memory store.
//
// Mirrors Python harness/memory.py, which maintains the lossless store backing
// the (lossy) kbinfos.Chunks list that feeds the LLM. Retrieval narrows chunks
// to the sentences that answer the current query, so the raw text has to be
// kept somewhere: a later gap query often needs a fact the earlier narrowing
// already threw away.
//
// The store lives on Kbinfos.Memory so it travels with the request.

const (
	// grepMaxChunks is Python memory.grep's default cap (_GREP_MAX_CHUNKS = 6).
	// MemoryGrep takes the limit explicitly and does NOT substitute this for a
	// non-positive value (Python has no such guard), so callers that want the
	// Python default pass it.
	grepMaxChunks = 6
	// grepMaxSentences caps sentences kept per chunk (hit + context).
	grepMaxSentences = 4
	// grepContextChars is the per-side char budget when expanding context.
	grepContextChars = 400
	// shortChunkChars: chunks at or below this length are kept whole, since
	// answers often live in short chunks.
	shortChunkChars = 200

	// MemorySearch tuning — mirror Python memory.search defaults.
	memoryDefaultTopN = 6
	// memoryMinRatio: a chunk is relevant when it shares >= 1 term AND >= this
	// fraction of the query's significant terms (normalized overlap bar so CN /
	// EN queries behave alike). Python's min_overlap param is declared but
	// unused in search(), so only the ratio matters.
	memoryMinRatio = 0.12
	// memoryMaxTerms caps how many significant terms we extract from a query.
	memoryMaxTerms = 18
)

var (
	// reTermPunct strips leading/trailing punctuation before escaping.
	reTermPunct = regexp.MustCompile(`^[\s.,:;!?'"()\[\]{}]+|[\s.,:;!?'"()\[\]{}]+$`)
	// reCJK detects CJK / kana / hangul, which get no \b anchor (a word boundary
	// never matches between CJK characters).
	reCJK = regexp.MustCompile(`[\x{4e00}-\x{9fff}\x{3040}-\x{30ff}\x{ac00}-\x{d7af}]`)
	// reCJKRun matches a maximal run of CJK/kana/hangul for 3-gram splitting.
	reCJKRun = regexp.MustCompile(`[\x{4e00}-\x{9fff}\x{3040}-\x{30ff}\x{ac00}-\x{d7af}]+`)
	// reLatinNum matches a maximal alphanumeric run (Latin words + numbers).
	reLatinNum = regexp.MustCompile(`[A-Za-z0-9]+`)
	// reDigits matches a pure digit run.
	reDigits = regexp.MustCompile(`\d+`)
)

// memoryStopwords mirrors Python memory._STOPWORDS: dropped from Latin query
// terms so "what / the / is" style words do not dominate the overlap score.
var memoryStopwords = map[string]struct{}{
	"what": {}, "which": {}, "how": {}, "many": {}, "much": {}, "does": {},
	"did": {}, "do": {}, "the": {}, "a": {}, "an": {}, "is": {}, "are": {},
	"was": {}, "were": {}, "be": {}, "been": {}, "being": {}, "of": {},
	"for": {}, "to": {}, "in": {}, "on": {}, "with": {}, "and": {}, "or": {},
	"by": {}, "from": {}, "at": {}, "it": {}, "its": {}, "this": {}, "that": {},
	"these": {}, "those": {}, "who": {}, "when": {}, "where": {}, "why": {},
	"than": {}, "then": {}, "there": {}, "their": {}, "they": {}, "them": {},
	"his": {}, "her": {}, "him": {}, "she": {}, "he": {}, "we": {}, "you": {},
	"your": {},
}

// MemoryAdd mirrors Python memory.add: merges raw retrieved chunks into the
// central store, losslessly, skipping chunks already present and those with no
// text.
func MemoryAdd(kb *Kbinfos, chunks []map[string]any) {
	if kb == nil || len(chunks) == 0 {
		return
	}
	seen := make(map[string]struct{}, len(kb.Memory))
	for _, c := range kb.Memory {
		seen[chunkKey(c)] = struct{}{}
	}
	added := 0
	for _, c := range chunks {
		if c == nil || chunkText(c) == "" {
			continue
		}
		k := chunkKey(c)
		if _, dup := seen[k]; dup {
			continue
		}
		seen[k] = struct{}{}
		kb.Memory = append(kb.Memory, c)
		added++
	}
	if added > 0 {
		_LOG.Printf("[Memory] stored %d new raw chunk(s); memory now has %d.", added, len(kb.Memory))
	}
}

// MemorySize mirrors Python memory.size.
func MemorySize(kb *Kbinfos) int {
	if kb == nil {
		return 0
	}
	return len(kb.Memory)
}

// MemoryClear mirrors Python memory.clear.
func MemoryClear(kb *Kbinfos) {
	if kb != nil {
		kb.Memory = nil
	}
}

// MemoryGrep mirrors Python memory.grep: returns memory chunks containing any of
// terms, narrowed to the matching sentence plus a small context window.
//
// terms are plain strings (entities / numbers / key phrases) as emitted by the
// analysis LLM. Each returned chunk carries a narrowed "content" so the caller
// can splice it straight into an evidence list. Empty on no-hit / no-memory.
//
// limit is the maximum number of chunks returned and is NOT normalized: Python's
// grep has no such guard, so a limit <= 0 makes the `len(hits) >= limit` check
// fire on the first hit (at most one chunk comes back). Callers wanting Python's
// default pass grepMaxChunks.
func MemoryGrep(kb *Kbinfos, terms []string, limit int) []map[string]any {
	if kb == nil || len(kb.Memory) == 0 || len(terms) == 0 {
		return nil
	}
	patterns, prefixPatterns := compileTerms(terms)
	if len(patterns) == 0 && len(prefixPatterns) == 0 {
		return nil
	}
	match := func(text string) bool {
		for _, p := range patterns {
			if p.MatchString(text) {
				return true
			}
		}
		// Prefix fallback: morphological tolerance. A gap term often differs from
		// the chunk's word by a suffix (gap "abbreviation" vs chunk "abbreviated").
		// Matching the leading stem at the START of a word lets a shared root hit.
		for _, p := range prefixPatterns {
			if p.MatchString(text) {
				return true
			}
		}
		return false
	}

	var hits []map[string]any
	for _, c := range kb.Memory {
		text := chunkText(c)
		if len(text) <= shortChunkChars {
			// Short chunk: keep whole (its answer may live anywhere in it).
			if match(text) {
				hits = append(hits, map[string]any{
					"content":  text,
					"doc_id":   c["doc_id"],
					"chunk_id": c["chunk_id"],
				})
				// This branch continues past the shared cap below, so enforce the
				// limit here too — otherwise a memory store full of matching short
				// chunks comes back whole. Python's memory.grep enforces it in the
				// same place.
				if len(hits) >= limit {
					break
				}
			}
			continue
		}
		sents := SplitSentences(text)
		var kept []string
		for i, s := range sents {
			if match(s) {
				for _, w := range sentenceSpanWindow(sents, i) {
					if !containsStr(kept, w) {
						kept = append(kept, w)
					}
				}
			}
			if len(kept) >= grepMaxSentences {
				break
			}
		}
		if len(kept) > 0 {
			hits = append(hits, map[string]any{
				"content":  strings.Join(kept, "\n"),
				"doc_id":   c["doc_id"],
				"chunk_id": c["chunk_id"],
			})
		}
		if len(hits) >= limit {
			break
		}
	}
	return hits
}

// compileTerms mirrors Python memory.grep's pattern construction: the escaped
// term (word-anchored when it is long enough and wholly alphanumeric) plus the
// leading-stem prefix pattern used as a fallback.
func compileTerms(terms []string) (patterns, prefixPatterns []*regexp.Regexp) {
	for _, t := range terms {
		if frag := escapeTerm(t); frag != "" {
			if p, err := regexp.Compile("(?i)" + frag); err == nil {
				patterns = append(patterns, p)
			}
		}
		stripped := strings.TrimSpace(t)
		prefix := ""
		if len([]rune(stripped)) >= 6 {
			prefix = string([]rune(stripped)[:5])
		}
		if prefix != "" && !reCJK.MatchString(prefix) {
			if p, err := regexp.Compile(`(?i)\b` + regexp.QuoteMeta(prefix)); err == nil {
				prefixPatterns = append(prefixPatterns, p)
			}
		}
	}
	return patterns, prefixPatterns
}

// escapeTerm mirrors Python _escape_term: strip surrounding punctuation, escape
// regex metacharacters, then anchor on word boundaries — except for CJK, where
// a \b anchor would never match.
func escapeTerm(term string) string {
	t := strings.TrimSpace(term)
	if t == "" {
		return ""
	}
	t = reTermPunct.ReplaceAllString(t, "")
	if t == "" {
		return ""
	}
	escaped := regexp.QuoteMeta(t)
	if reCJK.MatchString(t) {
		return escaped
	}
	rs := []rune(t)
	if len(rs) >= 3 && isAlnum(rs[0]) && isAlnum(rs[len(rs)-1]) {
		return `\b` + escaped + `\b`
	}
	return escaped
}

// sentenceSpanWindow mirrors Python _sentence_span_window: the hit sentence plus
// up to one neighbour on each side, clamped to a total length.
func sentenceSpanWindow(sents []string, idx int) []string {
	if idx < 0 || idx >= len(sents) {
		return nil
	}
	lo, hi := max(0, idx-1), min(len(sents), idx+2)
	total := 0
	var kept []string
	for _, s := range sents[lo:hi] {
		total += len(s)
		if total > grepContextChars*2 {
			break
		}
		kept = append(kept, s)
	}
	if len(kept) == 0 {
		return []string{sents[idx]}
	}
	return kept
}

func isAlnum(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsDigit(r)
}

func containsStr(ss []string, s string) bool {
	for _, v := range ss {
		if v == s {
			return true
		}
	}
	return false
}

// MemorySearch mirrors Python memory.search: relevance-ranked retrieval over the
// raw-chunk memory store (a retrieval-reuse cache, NOT a noise-injection source).
// Unlike MemoryGrep (loose keyword hit) it keeps only chunks whose overlap with the
// query's SIGNIFICANT terms clears a normalized bar, so a fact retrieved earlier can
// be reused instead of re-querying the index.
//
// Language-agnostic term extraction (mirrors Python _significant_terms):
//   - numbers kept verbatim;
//   - CJK runs split into character 3-grams (no word boundaries exist);
//   - Latin alphanumeric runs lowercased, stopword-filtered, len >= 3.
//
// A chunk is relevant when it shares >= 1 term AND >= minRatio of the query's
// significant terms; results are ranked by hit count then text length, capped at
// topN. Returns nil when nothing clears the bar (the caller falls back to a
// knowledge-base search). minRatio <= 0 falls back to memoryMinRatio.
func MemorySearch(kb *Kbinfos, query string, topN int, minRatio float64) []map[string]any {
	if kb == nil || len(kb.Memory) == 0 {
		return nil
	}
	if topN <= 0 {
		topN = memoryDefaultTopN
	}
	if minRatio <= 0 {
		minRatio = memoryMinRatio
	}
	terms := significantTerms(query)
	if len(terms) == 0 {
		return nil
	}
	matchers := buildTermMatchers(terms)
	n := len(terms)

	type scoredChunk struct {
		hits int
		text string
		c    map[string]any
	}
	var scored []scoredChunk
	for _, c := range kb.Memory {
		text := chunkText(c)
		if text == "" {
			continue
		}
		hits := 0
		for _, m := range matchers {
			if m.matches(text) {
				hits++
			}
		}
		if hits < 1 {
			continue
		}
		if float64(hits)/float64(n) < minRatio {
			continue
		}
		scored = append(scored, scoredChunk{hits: hits, text: text, c: c})
	}
	if len(scored) == 0 {
		return nil
	}
	// Rank by hit count desc, then text length desc (Python: (-hits, -len)).
	sort.SliceStable(scored, func(i, j int) bool {
		if scored[i].hits != scored[j].hits {
			return scored[i].hits > scored[j].hits
		}
		return len(scored[i].text) > len(scored[j].text)
	})
	if len(scored) > topN {
		scored = scored[:topN]
	}
	out := make([]map[string]any, 0, len(scored))
	for _, s := range scored {
		out = append(out, map[string]any{
			"content":    s.text,
			"doc_id":     s.c["doc_id"],
			"chunk_id":   s.c["chunk_id"],
			"similarity": float64(s.hits),
		})
	}
	rq := []rune(query)
	if len(rq) > 60 {
		rq = rq[:60]
	}
	_LOG.Printf("[Memory.search] query=%q -> %d relevant chunk(s) (ratio>=%.2f, %d terms)", string(rq), len(out), minRatio, n)
	return out
}

// significantTerms mirrors Python memory._significant_terms: language-agnostic
// significant-term extraction, de-duplicated and capped at memoryMaxTerms.
func significantTerms(text string) []string {
	var out []string
	seen := make(map[string]struct{})
	push := func(tok string) {
		if tok == "" {
			return
		}
		if _, dup := seen[tok]; dup {
			return
		}
		seen[tok] = struct{}{}
		out = append(out, tok)
	}
	// Numbers anywhere.
	for _, m := range reDigits.FindAllString(text, -1) {
		push(m)
		if len(out) >= memoryMaxTerms {
			return out
		}
	}
	// CJK runs -> 3-grams (and the whole run if shorter than 3).
	for _, run := range reCJKRun.FindAllString(text, -1) {
		rs := []rune(run)
		if len(rs) < 3 {
			push(run)
		} else {
			for i := 0; i+3 <= len(rs); i++ {
				push(string(rs[i : i+3]))
			}
		}
		if len(out) >= memoryMaxTerms {
			return out
		}
	}
	// Latin words (stopword-filtered, len >= 3); pure-digit runs already handled.
	for _, m := range reLatinNum.FindAllString(text, -1) {
		if isAllDigits(m) {
			continue
		}
		low := strings.ToLower(m)
		if len(low) >= 3 {
			if _, stop := memoryStopwords[low]; !stop {
				push(low)
			}
		}
		if len(out) >= memoryMaxTerms {
			return out
		}
	}
	return out
}

// termMatcher precompiles one query term's match predicate so MemorySearch
// does not recompile regexes per chunk.
type termMatcher struct {
	// kind: 0 = CJK substring, 1 = digit substring, 2 = Latin word-boundary
	// (with a short prefix fallback for inflectional variants).
	kind int
	sub  string
	re   *regexp.Regexp
	pre  *regexp.Regexp
}

func buildTermMatchers(terms []string) []termMatcher {
	matchers := make([]termMatcher, 0, len(terms))
	for _, t := range terms {
		if run := reCJKRun.FindString(t); run != "" && run == t {
			// Pure CJK term: literal substring (no word boundary).
			matchers = append(matchers, termMatcher{kind: 0, sub: t})
			continue
		}
		if isAllDigits(t) {
			matchers = append(matchers, termMatcher{kind: 1, sub: t})
			continue
		}
		rs := []rune(t)
		re, _ := regexp.Compile(`(?i)\b` + regexp.QuoteMeta(t) + `\b`)
		var pre *regexp.Regexp
		if len(rs) >= 6 {
			pre, _ = regexp.Compile(`(?i)\b` + regexp.QuoteMeta(string(rs[:5])))
		}
		matchers = append(matchers, termMatcher{kind: 2, re: re, pre: pre})
	}
	return matchers
}

func (m termMatcher) matches(text string) bool {
	switch m.kind {
	case 0, 1:
		return strings.Contains(text, m.sub)
	default:
		if m.re != nil && m.re.MatchString(text) {
			return true
		}
		if m.pre != nil && m.pre.MatchString(text) {
			return true
		}
		return false
	}
}

// IsStopword reports whether w is one of Python memory._STOPWORDS.
//
// Go's fan-out needs it because Python aliases the same set as
// _FANOUT_STOPWORDS (agentic_rag_graph.py:_expand_fanouts) when filtering candidate terms.
func IsStopword(w string) bool {
	_, ok := memoryStopwords[w]
	return ok
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}
